File: files/CH-164.R
library(tidyverse)
library(readxl)

path = "files/CH-164 Extract from Text.xlsx"
input = read_excel(path, range = "B2:C7")
test  = read_excel(path, range = "E2:F7")

maxDepth = function(S) {
  get_chars = function(str) {
    tibble(pos = 1:nchar(str), char = strsplit(str, "")[[1]])
  }
  
  check = if_else(str_detect(S, "^[{]+[0-9,]+[}]+$"), -1, 0)
  df = get_chars(S)
  
  df = df %>%
    mutate(count = ifelse(char == "{", 1, ifelse(char == "}", -1, 0)),
           cum_sum = cumsum(count))
  
  max_depth = max(df$cum_sum) + check
  
  return(max_depth)
}

result = input %>%
  mutate(Depth = map_dbl(Value, maxDepth))

all.equal(result$Depth, test$Depth)
#> [1] TRUE

File: files/100-199/164/CH-164.py
import pandas as pd
import re

path = "CH-164 Extract from Text.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=6)
test = pd.read_excel(path, usecols="E:F", skiprows=1, nrows=6)

def max_depth(S):
    get_chars = lambda string: pd.DataFrame({'pos': range(1, len(string) + 1), 'char': list(string)})
    
    check = -1 if re.match(r'^[{]+[0-9,]+[}]+$', S) else 0
    df = get_chars(S)
    
    df['cum_sum'] = df['char'].apply(lambda x: 1 if x == '{' else (-1 if x == '}' else 0)).cumsum()

    max_depth = df['cum_sum'].max() + check
    
    return max_depth

input['depth'] = input['Value'].apply(max_depth)

print(input['depth'].equals(test['Depth'])) # True